### PR TITLE
explore/formly-integration

### PR DESCRIPTION
**Pull request by MaxKless**
## PR Type

[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no API changes)
[ ] Build-related changes
[ ] CI-related changes
[ ] Documentation content changes
[ ] Application / infrastructure changes
[ ] Other: 

### Done
- [x] Docs - specific templateOptions
- [x] Final rebase and commit structure
- [x] address model typing as Address / Partial<Address>
- [x] position of validation icons on select boxes (cross overlaps arrow)
- [x] e2e tests
- [x] rename and adapt FormlyTesting2Module
- [x] new fields tests
- [x] full registration page
- [x] email field
- [x] password field
- [x] Add Taxation IDs to configs
- [x] mark old components as deprecated
- [x] forms-dynamic module dead code / deprecation ?
- [x] split commits to framework / implementation
- [x] Checkout address form
- [x] Account address form
- [x] tslint rules
- [x] seperate b2b and b2c
- [x] reevaluate grouping of fields vs styled margins -> any upside to grouping them?
- [x] payment formly (mapped from server)
- [x] write form configuration schematic tests
- [x] adapt addTokenProviderToNgModule with automatic relative path construction
- [x] move default values to template components
- [x] remove country / region from formly service
- [x] properly integration FormlyTestingModule or change to forTesting() function
- [x] refactor tests to use testing module
- [x] fix country select only appears after click
- [x] write address form tests
- [x] red required star
- [x] clarify ux on submission validation -> select different country
- [x] finalize each form config & validation
- [x] write form configuration schematic
- [x] registration form adjustment
- [x] rename custom-xx-field to xx-field / ish-xx-field
- [x] unit tests -- reevaluate
- [x] fix required star

## Does this PR Introduce a Breaking Change?

[ ] Yes
[ ] No

### CODE REVIEW

**Review by M-Behr — CHANGES_REQUESTED** [review]
Good work! :) For the most part, it works as before.

I found some issues:
- [ ] only country is marked as required field on address form (e.g. registration page, my account, checkout)
- [ ] font size of error messages and position of validation icons on select boxes (cross overlaps arrow)
- [ ] display first form hint and below error message (e.g. select United States and see "APO/FPO" link) => Your order is also possible but if we rework it step by step, I would like to avoid inconsistencies. Please stay with the current order.
FormValidation

- [ ] contact us form is not responsive => please align labels top
- [ ] form hint for comment "Your message: 30000 characters remaining." is missing 
- [ ] submit button is not in line with the form fields

- [ ] add address in checkout does not work => error message: customer.address_not_found.error => I think the functionality tries to change the selected address instead of create a new address.
**Current behavior:** click "add address" link, enter a new address and click "save changes" => error
**Expected:** click "add address" link, enter a new address and click "save address" => address is created and selected as invoice or shipping address.

- [ ] Taxation ID is displayed twice on b2b registration form
- [ ] Taxation ID is part of address template in my account/checkout?

I think we need to achieve a very flexible address template. I noted down that we separated the registration page in 3 sections: Company Information, Personal Information and Address.
For the address template in my account/checkout, we use company name 1 and 2 from company information and the address component. I created this page after our last discussion with the team:
Address Template Rework

**Review by dhhyi — CHANGES_REQUESTED** [review]
Part 1:
- mainly focused on contact-us
Later:
- duplications in adressform models could be reduced. I'll propose a suggestion.

**Review by M-Behr — CHANGES_REQUESTED** [review]
Hi Max,
your integration looks pretty good.

I found a little little thing in the registration form validation:
- [ ] Please display the error message in a new line (not next to the hint).
2021-02-16 10_32_20-Register _ Intershop PWA

And I found some crazy behavior on the address page in checkout. It took me a little more research to figure out why this was happening.
- [ ] Edit an address in the checkout. Save the changes and the edited address is displayed. All is fine. If you now click again the edit icon an empty address form is displayed and for the add address an prefilled address form is displayed. The forms are mixed up.

**Review by iwiederhold — DISMISSED** [review]

**Review by marschmidt89 — DISMISSED** [review]
Documentation review done.

**Review by dhhyi — CHANGES_REQUESTED** [review]

File: src/app/shared/formly-address-forms/configurations/default/address-form-default.configuration.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+      ...(this.businessCustomer ? businessModelPartial : {}),
+      firstName: model.firstName ?? '',
+      lastName: model.lastName ?? '',
+      addressLine1: model.addressLine1 ?? '',
+      addressLine2: model.addressLine2 ?? '',
+      postalCode: model.postalCode ?? '',
+      city: model.city ?? '',
+      mainDivisionCode: model.mainDivisionCode ?? '',
+      phoneHome: model.phoneHome ?? '',
+    };
```
[the remark this answers is not in the document]
Comment (dhhyi) [reply src/app/shared/formly-address-forms/configurations/default/address-form-default.configuration.ts:36]: why even pick 'mainDivisionCode'? It should be unusable, if the country changed 🤔

**Review by dhhyi — DISMISSED** [review]

**Review by M-Behr — APPROVED** [review]

**Review by SGrueber — APPROVED** [review]

**Pull request merged**